### DISCUSSION TITLE
How to decrypt password in nodejs?

### DISCUSSION THREAD

**Discussion by IRediTOTO** [dani-garcia/vaultwarden#4597] [Q&A]
Hello, Im trying to use Vaultwarden as a service to store users sensitive info but I don't know how to decrypt password from key, private key, access_token,...
Can you show me the way to decrypt it? Yes I googled `https://github.com/GurpreetKang/BitwardenDecrypt` but it not help me how to decrypt because not familiar with the py.
Thank you for any help!!!

**Comment by stefan0xC** [comment]
What do you mean by "how to decrypt password from key, private key, access_token,..."? I'd recommend reading the [Bitwarden Security White Paper](https://bitwarden.com/help/bitwarden-security-white-paper/) to understand how the data is encrypted.

**Reply by IRediTOTO** [reply]
I just want to right a js function to decrypt the cipher string. Can you give a look to my code?  Thank you for any help <3

```

export async function GET(request) {
    const decryptedText = decryptCipherString(
        cipherString,
        masterPassword,
        email,
    );

    return Response.json({
        decryptedText,
    });
}

// PBKDF2 key derivation
const deriveKey = (password: string, salt: Buffer): Buffer => {
    const iterations = 600000;
    return crypto.pbkdf2Sync(password, salt, iterations, 32, "sha256");
};

// Decrypt function
const decryptCipherString = (
    cipherString: string,
    masterPassword: string,
    encryptedKey: string,
) => {
    // Decode components
    const iv = Buffer.from(cipherString.split(".")[1].split("|")[0], "base64");
    const encryptedData = Buffer.from(
        cipherString.split(".")[1].split("|")[1],
        "base64",
    );
    const mac = Buffer.from(cipherString.split(".")[1].split("|")[2], "base64");

    // Derive the key
    const key = deriveKey(masterPassword, Buffer.from(encryptedKey, "base64"));

    // Verify the MAC for AES-CBC (encryptionType 2)
    const computedMac = crypto
        .createHmac("sha256", key)
        .update(iv)
        .update(encryptedData)
        .digest();
    if (!crypto.timingSafeEqual(computedMac, mac)) {
        throw new Error("MAC verification failed");
    }

    // Decrypt the data
    const decipher = crypto.createDecipheriv("aes-256-cbc", key, iv);
    let decrypted = decipher.update(encryptedData);
    decrypted = Buffer.concat([decrypted, decipher.final()]);

    return decrypted.toString();
};
```

**Reply by BlackDex** [reply]
Not sure you are at the right address here to ask this question though. Vaultwarden it self does nothing at all in regards to encrypting and decrypting vault items. If you want any useful help you might try the Bitwarden Community forums. https://community.bitwarden.com/
